compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Cleanup.cpp

```diff
@@ -0,0 +1,147 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/Transforms/PassDetail.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree/compiler/Codegen/Common/Transforms.h"
+#include "iree/compiler/Codegen/Transforms/Transforms.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/TilingInterfaceImpl.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Patterns.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/LoopInvariantCodeMotionUtils.h"
+
+#define DEBUG_TYPE "iree-cleanup"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+static void loopIndependentCodeMotion(Operation *funcOp, IRRewriter &rewriter) {
+  // This assumes LICM never removes operations so we don't need tracking.
+  // TODO: confirm / revisit this assumption and plumb a rewriter through
+  // upstream moveLoopInvariantCode if necessary.
+  funcOp->walk([](LoopLikeOpInterface loopLike) {
+    // Do not hoist from scf.forall ops. These capture isolated computations
+    // that will be mapped to a certain level in the GPU hierarchy (e.g.,
+    // GPU blocks), so hoisting is not desired.
+    if (!isa<scf::ForallOp>(loopLike.getOperation()))
+      moveLoopInvariantCode(loopLike);
+  });
+  // For now, put single loop promotion as part of licm. Underlying
+  // implementations perform splice operations which shouldn't need
+  // tracking.
+  // TODO: confirm / revisit this assumption and plumb a rewriter through
+  // upstream moveLoopInvariantCode if necessary.
+  funcOp->walk([&](Operation *op) {
+    (void)llvm::TypeSwitch<Operation *, LogicalResult>(op)
+        .Case<affine::AffineForOp, scf::ForOp>(
+            [&](auto loop) { return loop.promoteIfSingleIteration(rewriter); })
+        .Default([](Operation *) { return success(); });
+  });
+}
+
+/// COPIED FROM IREE TRANSFORM.
+/// Fold `tensor.pad(cst, tensor.extract*(linalg.fill(cst)))` into
+/// `linalg.fill(cst, empty)` when the padding constant and the fill constant
+/// are the same.
+/// This seems generally desirable as a folding but may be too intrusive, so we
+/// only apply it selectively for now.
+// TODO: atm hardcoded on linalg.fill but we could take any result of any
+// generic that yields a constant in that result.
+struct FoldFillIntoPad : public OpRewritePattern<tensor::PadOp> {
```

Comment:
Isnt there a pattern to do this upstream already?

Reply:
Found it. Here https://github.com/llvm/llvm-project/blob/9ed30012fb4f43de42ef2f265fe384d9d0b0edf2/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L601